odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self.left.derivative(point)
+
+
+class TranslatedFunctional(Functional):
+
+    """Implementation of the translated functional.
+
+    Given a functional ``f`` and an element ``translation`` in the domain of
+    ``f``, this corresponds to the functional ``f(. - translation)``.
+    """
+
+    def __init__(self, func, translation):
+        """Initialize a new instance.
+
+        Given a functional ``f(.)`` and a vector ``translation`` in the domain
+        of ``f``, this corresponds to the functional ``f(. - y)``.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional which is to be translated.
+        translation : `LinearSpaceVector`
+            Element in ``func.domain``, with which the argument is translated.
+        """
+
+        if not isinstance(func, Functional):
+            raise TypeError('`func` {!r} not a `Functional` instance'
+                            ''.format(func))
+
+        if translation not in func.domain:
+            raise TypeError('`translation` {!r} not in func.domain {!r}'
+                            ''.format(translation.space, func.domain))
+
+        super().__init__(domain=func.domain, linear=False,
+                         grad_lipschitz=func.grad_lipschitz)
+
+    # TODO: Add case if we have translation -> scaling -> translation?
+        if isinstance(func, TranslatedFunctional):
+            self._original_func = func.original_func
+            self._translation = func.translation + translation
+
+        else:
+            self._original_func = func
+            self._translation = translation
+
+    @property
+    def original_func(self):
+        return self._original_func
+
+    @property
+    def translation(self):
+        return self._translation
+
+    def _call(self, x):
+        """Evaluates the functional in a point ``x``.
+
+        Parameters
+        ----------
+        x : `LinearSpaceVector`
+            Element in the domain of the functional.
+            The point in which the functional is evaluated
+
+        Returns
+        -------
+        `self(x)` : `float`
+            Evaluation of the functional, which is a constant.
+        """
+        return self._original_func(x - self._translation)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        The operator is given by a transation of the gradient operator of the
+        original functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+
+        class TranslatedGradientOperator(Operator):
```

Comment:
Didn't we add it so that `Operator + vector == Operator(x) + vector`? In this case we want `grad_(x - translation)`.

One could do
```python
return self.original_func.gradient * ResidualOperator(IdentityOperator(self.domain), self.translation)
```
but is it worth it?

Any other rewritings? Otherwise I keep it as it is.